Support printing pids in Logger metadata.

MSch opened the pull request:
Fixes #3336.

This _only_ supports printing metadata that conforms to `String.Chars` or is a pid. I'm not sure if that's what we want, I'd like to also be able to have refs as metadata for example.

Also please backport to 1.0

josevalim commented:
What about printing it only as `pid=0.10.3`?

MSch commented:
Right, forgot how Elixir inspect works... I'd prefer `pid=<0.13.0>` over `pid=0.13.0`, since I imagine setting e.g. `link` to the process calling `link_me` in Ecto's worker. It'd be clearer that this is a pid then IMO.

What about references? (We use references to identify some operations, so adding them easily to the metadata would be nice)

josevalim commented:
Yeah, we can support both.

Well, "pid" is the name of the tag. So if you have "key: self()" as
metadata, it will show up as key=0.13.0. The idea is that you know what key
is, so you don't need to pollute the log with duplicated information. That
said, key=0.13.0  or key=<0.13.0> are both fine by me.

josevalim commented on lib/logger/lib/logger/formatter.ex:
```diff
@@ -118,4 +118,14 @@ defmodule Logger.Formatter do
   defp levelpad(:error), do: ""
+
+  defp metadata(pid) when is_pid(pid) do
+    "#PID" <> string = inspect(pid)
```
:erlang.pid_to_list

josevalim commented on lib/logger/lib/logger/formatter.ex:
```diff
@@ -118,4 +118,14 @@ defmodule Logger.Formatter do
+    string
+  end
+  defp metadata(pid) when is_reference(pid) do
+    "#Reference" <> string = inspect(pid)
```
:erlang.ref_to_list

MSch commented:
Fixed and added support for refs.

My thoughts are that if the SQL logs in e.g. phoenix are having a metadata key `link` it'll be (more) obvious what  that is if it's surrounded by `<` and `>` and the phoenix request handling process' pid key is equal to it.

MSch commented:
@josevalim fixed. Stripping `#Ref` ok with you? I thought since refs have 4 parts and pids have 3 it's fine.
